validator/validator.go

```diff
@@ -0,0 +1,82 @@
+// Package validator provides functionality for validating groups and units.
+package validator
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/giantswarm/inago/controller"
+)
+
+// ValidateRequest takes a Request, and returns whether it is valid or not.
+// If the request is not valid, the error provides more details.
+func ValidateRequest(request controller.Request) (bool, error) {
+	// Check there are units in the group.
+	if len(request.Units) == 0 {
+		return false, noUnitsInGroupError
+	}
+
+	// Check that each unit name is prefixed with the group name.
+	for _, unit := range request.Units {
+		if !strings.HasPrefix(unit.Name, request.Group) {
+			return false, badUnitPrefixError
+		}
+	}
+
+	// Check that all units either have @ or they don't.
+	numUnitsWithAtSymbol := 0
+	for _, unit := range request.Units {
+		if strings.Contains(unit.Name, "@.") {
+			numUnitsWithAtSymbol++
+		}
+	}
+	if numUnitsWithAtSymbol > 0 && numUnitsWithAtSymbol < len(request.Units) {
+		return false, mixedSliceInstanceError
+	}
+
+	// Test there are not multiple @ symbols in the group name.
+	if strings.Count(request.Group, "@") > 1 {
+		return false, multipleAtInGroupNameError
+	}
+
+	// Test there are not multiple @ symbols in any unit name.
+	for _, unit := range request.Units {
+		if strings.Count(unit.Name, "@") > 1 {
+			return false, multipleAtInUnitNameError
+		}
+	}
+
+	// Test that all unit names are unique
+	for i := 0; i < len(request.Units)-1; i++ {
```

Comment:
Yeah, I did it this way so I can check pairwise without panicing - I don't get how to do that with your syntax